File: python/sor.py
import sympy as sp
sp.init_printing(use_unicode = True)
def optimizeW(L, U, D):
    Tj = (D**-1)*(L+U) 
    EigTj = map(lambda x: float(abs(x).expand(complex = True)), list(Tj.eigenvals().keys()))
    TjRatio = max(EigTj)
    w = 2/(1 + sp.sqrt(1 - TjRatio**2))
    return w

def infinityNorm(A):
    B = list()
    for i in range(0, len(A.col(0))):
        B.append(sum(abs(A.row(i))))
    return max(B)
    
    
def sor(A, b, x_0, iter, tol):
    Anp = sp.Matrix(A)
    bnp = sp.Matrix(b)
    x_i = sp.Matrix(x_0)
    D = sp.diag(*Anp.diagonal())
    print("D = ")
    sp.pprint(D)
    print("\n")
    U = D - Anp.upper_triangular()
    print("U = ")
    sp.pprint(U)
    print("\n")
    L = D-Anp.lower_triangular()
    print("L = ")
    sp.pprint(L)
    print("\n")
    wCut = '{:.4f}'.format(optimizeW(L, U , D))
    w = float(wCut)
    Tw = ((D-w*L)**-1)*((1-w)*D + w*U)
    print("Tw = ")
    sp.pprint(Tw)
    print("\n")
    Cw = w*((D-w*L)**-1)*bnp
    print("Cw = ")
    sp.pprint(Cw)
    print("\n")
    print("spectral ratio:", w - 1)
    print("w = ", w)
    for i in range(0, iter):
        x = Tw*x_i + Cw
        if Anp*x.evalf(6) == bnp or infinityNorm(x-x_i)/infinityNorm(x) < tol:
            print(" X = ")
            sp.pprint(x.evalf(6))
            break 
        x_i = x
        print(f"x_{i+1} = ", list(x.evalf(6)))
A = eval(input("Ingrese la matriz: "))
b = eval(input("introduzca el vector b: "))
x_0 = eval(input("ingrese el vector inicial: "))
iter = int(input("ingrese el nùmero de iteraciones: "))
tol = float(input("Ingrese la tolerancia: "))
sor(A, b, x_0, iter, tol)

File: octave/sor.m
function p = sor(A, b, x_0, iter, tol)
  D = diag(diag(A));
  fprintf("D = \n")
  disp(D)
  L = D- tril(A);
  fprintf("L = \n")
  disp(L)
  U = D - triu(A);
  fprintf("U = \n")
  disp(U)
  fprintf("w más optimo: ")
  w = optimezeW(L, U, D);
  disp(w)
  for i=1:iter
    x = inv(D-w*L)*((1-w)*D + w*U)*x_0 + w*inv(D-w*L)*b
    if A*x - b == zeros(length(A), 1) | infNorm(x - x_0)/infNorm(x) < tol  
      break;
    end
    x_0 = x;
  end
  
end

function w = optimezeW(L,U,D)
  Tj = inv(D)*(L+U);
  Tjratio = max(abs(eig(Tj)));
  w = 2/(sqrt(1-(Tjratio)^2) + 1);
end

function n = infNorm(A)
  n = max(sum(abs(A)));
end
